sound/soc/sof/sof-audio.c

```diff
@@ -495,23 +485,179 @@ int sof_machine_register(struct snd_sof_dev *sdev, void *pdata)
 
 	/* register machine driver, pass machine info as pdata */
 	audio_data->pdev_mach =
-		platform_device_register_data(sdev->dev, drv_name,
+		platform_device_register_data(audio_data->dev, drv_name,
 					      PLATFORM_DEVID_NONE, mach, size);
 	if (IS_ERR(audio_data->pdev_mach))
 		return PTR_ERR(audio_data->pdev_mach);
 
-	dev_dbg(sdev->dev, "created machine %s\n",
+	dev_dbg(audio_data->dev, "created machine %s\n",
 		dev_name(&audio_data->pdev_mach->dev));
 
 	return 0;
 }
 EXPORT_SYMBOL(sof_machine_register);
 
-void sof_machine_unregister(struct snd_sof_dev *sdev, void *pdata)
+void sof_machine_unregister(void *data)
 {
-	struct snd_sof_audio_data *audio_data = sdev->sof_audio_data;
+	struct snd_sof_audio_data *audio_data =
+		(struct snd_sof_audio_data *)data;
 
 	if (!IS_ERR_OR_NULL(audio_data->pdev_mach))
 		platform_device_unregister(audio_data->pdev_mach);
 }
 EXPORT_SYMBOL(sof_machine_unregister);
+
+static int sof_destroy_pipelines(struct device *dev)
+{
+	struct snd_sof_audio_data *audio_data =	sof_get_client_data(dev);
+	struct snd_sof_widget *swidget;
+	struct sof_ipc_free ipc_free;
+	struct sof_ipc_reply reply;
+	int ret = 0;
+
+	list_for_each_entry_reverse(swidget, &audio_data->widget_list, list) {
+		memset(&ipc_free, 0, sizeof(ipc_free));
```

Comment:
put this 3 lines down - after skipping, but in fact I'd just remove it - you're assigning all the `ipc_free` fields below.